### PR TITLE
Burst-Fire Weapons

### PR DESCRIPTION
**Pull request by Jade-Harleyy**
This PR implements two types of burst-fire weapons:
- **Save**: The burst count will not reset when the trigger is released.
- **Forced**: The burst will continue even if the trigger is released.

They can be added to a RangedWeapon component using the following properties:
- **BurstMode** (BurstType) - The type of burst the weapon performs.
- **BurstCount** (int) - The amount of shots fired in a single burst.
- **BurstReload** (float) - How long the user has to wait before they can fire the weapon again after a burst (in seconds).
- **BurstReloadNoSkill** (float) - Burst reload time at 0 skill level. Reload time scales with skill level up to the Weapons skill requirement.

### CODE REVIEW

File: Barotrauma/BarotraumaShared/SharedSource/Items/Components/Holdable/RangedWeapon.cs [hunk 1]
```diff
@@ -138,6 +183,9 @@ public Vector2 TransformedBarrelPos
         private float currentChargeTime;
         private bool tryingToCharge;
 
+        private Character prevUser;
+        private int burstPos;
```
Comment (Regalis11) [inline Barotrauma/BarotraumaShared/SharedSource/Items/Components/Holdable/RangedWeapon.cs:187]: Not clear what this means, needs a more descriptive name and/or comment. I think something like `burstCounter` or `burstIndex` would be more clear than `burstPos`.

File: Barotrauma/BarotraumaShared/SharedSource/Items/Components/Holdable/RangedWeapon.cs [hunk 2]
```diff
@@ -233,34 +305,44 @@ private static float ApplyDualWieldPenaltyReduction(Character character, float o
...
+
+            ReloadTimer = newReload / (1 + character?.GetStatValue(StatTypes.RangedAttackSpeed) ?? 0f);
             ReloadTimer /= 1f + item.GetQualityModifier(Quality.StatType.FiringRateMultiplier);
+        }
 
+        private readonly List<Body> ignoredBodies = new List<Body>();
+        public override bool Use(float deltaTime, Character character = null)
+        {
+            if (character == null || character.Removed) return false;
+            if (ReloadTimer > 0) return false;
```
Comment (Regalis11) [inline Barotrauma/BarotraumaShared/SharedSource/Items/Components/Holdable/RangedWeapon.cs:334]: Curved brackets should be used in single-line if clauses too.

**Review by Regalis11 — CHANGES_REQUESTED** [review]
Thank you for the contribution! I found some minor code quality issues that I feel should be addressed before I'm willing to approve of the PR.

**Review by Regalis11 — CHANGES_REQUESTED** [review]
We took a look at this internally, and it seems I was a bit too hasty in approving this. I feel this will need quite a bit more work, and testing that all of it works as intended seems potentially so time-consuming that I unfortunately can't make any promises that we'd including this in the game even if all the issues found so far were addressed:

- The main issue is that `burstIndex` is not synced in multiplayer. If a clients' index is out of sync for whatever reason, for example if they join mid-round after the weapon has been fired, or if they miss a single shot due to networking issues, they'll reload timers will be completely off, leading to even worse syncing issues (e.g. the weapon going off client-side, but not server-side because the longer reload timer was active, or vice versa).
- It's impossible to create a burst-fire weapon that allows firing rapid single shots with this system. I.e. if the weapon is set to fire in bursts of three, if you fire 3 individual shots, you'll need to wait for the "longer" reload time. This doesn't feel that good to use: there's very little feedback to the player, so in practice it often just feels the gun doesn't fire for some reason. This also means weapons like the deadeye carbine cannot be implemented with this new system, which means we'd need to keep it unchanged, and in my opinion it's not a good thing if there are burst fire weapons implemented using two different systems (whose behavior is subtly different).
- There were some comments about the names of the burst fire modes being confusing, and in hindsight, I agree: if you look at an XML and see that a weapon's burst fire mode is "save", I think it'd be really hard to tell what it means unless you're familiar with how the feature works technically (that it "saves" the index of the burst).
- It's not obvious that the `BurstReload` and `BurstReloadNoSkill` values are affected by a separate property, not the "normal" weapons skill requirement that affects the normal reload. I don't quite see why a separate property is needed, but if it is, at the very least there should be a warning if the skill-based reload times have been configured without configuring the skill requirement.

**Comment by Jade-Harleyy** [comment]
> It's not obvious that the `BurstReload` and `BurstReloadNoSkill` values are affected by a separate property, not the "normal" weapons skill requirement that affects the normal reload.

I'm not sure what you mean by this; `ReloadNoSkill` and `BurstReloadNoSkill` are both still dependent on `ReloadSkillRequirement`.

**Review by Regalis11 — CHANGES_REQUESTED** [review]
Syncing the burst index doesn't work. There's no logic that would create any events for it. I also believe this approach would not work even if the events were created: the clients must be able to launch the projectile with no delay (= use client-side prediction) for the weapons to feel responsive. The network syncing however always has a delay. 

What this mean the burst index would be "out of sync" client-side: 
1. A client fires a gun, sets the burst index to let's say 1
2. The server notices the client fired the gun and sets it to 1 too and informs clients it was set to 1
3. The client keeps firing, sets the burst index to 2
4. The client receives an update from the server, saying the index is now 1
5. The end result being that the client will probably fire too many shots per burst, depending on the timing of the events.

I think addressing this will be very complicated and take a lot of testing to get working reliably, so even if you were to find a way to get it working, I can't make any promises that we'd be including it in the game.

**Comment by Jade-Harleyy** [comment]
Even if the entire PR can't be implemented, I added a new property in the last commit that shouldn't have any problems, would be very useful, and could potentially be cherry-picked. :)

**Review by Regalis11 — CHANGES_REQUESTED** [review]
The syncing issue I described in the previous review has not been addressed. I believe something like client-side prediction of the burst index with delayed corrections is needed to get it working properly.

**Comment by Regalis11** [comment]
Closing for the reasons explained in the above reviews: this does not seem to be usable in it's current form in multiplayer.

**Pull request closed without merging**